include/swift/AST/DiagnosticsSema.def

```diff
@@ -550,6 +550,9 @@ ERROR(expr_selector_property_not_settable,none,
       (DescriptiveDeclKind, DeclName))
 ERROR(expr_selector_property_setter_inaccessible,none,
       "setter of %0 %1 is inaccessible", (DescriptiveDeclKind, DeclName))
+ERROR(expr_selector_cannot_be_used,none,
+     "cannot use %0 as a selector due to associatedtype requirement on %1",
```

Comment:
Yeah, it might be tricky to check all those cases though. At the moment, it only checks for presence of `associatedtype` members, hence why the diagnostic mentions it. I guess I could reword it to something more specific, like `cannot use %0 as a selector because %1 is not exposed to Objective-C due to associatedtype requirements`?

Reply:
That's not correct, though. Most protocols aren't exposed to Objective-C even when they can be.

Reply:
Oh, I didn't know that! How about `cannot use %0 as a selector because it is not exposed to Objective-C due to associatedtype requirement on %1`? It's what `expr_selector_not_objc` says and I have just altered it a bit to include `associatedtype`.

Reply:
Hmm, if I don't add this check, then the fixit would be applied on the protocol & then you'll get another error: `Associated type 'Foo' cannot be declared inside '@objc' protocol 'Baz'`. 

I could reword it to something like `cannot use %0 as a selector because %1 is not exposed to Objective-C` & then add a note `It is not possible to expose a protocol to Objective-C if it has associatedtype requirements, non-bridged Swift types, etc`.  

Is this better? or do you have any refinements in mind that I could make to the diagnostic?

Reply:
Honestly, at this point I think it'd be better not to offer the fix-it at all. "cannot use %0 as a selector because protocol %1 is not exposed to Objective-C" is probably the best we can do. I already felt weird about offering a fix-it that might not actually work *and* that might not be what the user wants (such as in the original bug report, where the default implementation won't work because it can't be exposed to Objective-C).

Reply:
@jrose-apple Hmm seems like a reasonable approach. I have updated the code now!